floris/simulation/farm.py

```diff
@@ -83,7 +84,81 @@ class Farm(BaseClass):
+        # In other words, if the turbine type is already in the cache, skip that iteration of
+        # the for-loop.
+        turbine_definition_cache = {}
+        for t in self.turbine_type:
+            # If a turbine type is a dict, then it was either preprocessed by the yaml
+            # library to resolve the "!include" or it was set in a script as a dict. In either case,
+            # add an entry to the cache
+            if isinstance(t, dict):
+                if t["turbine_type"] in turbine_definition_cache:
+                    continue
+                turbine_definition_cache[t["turbine_type"]] = t
+
+            # If a turbine type is a string, then it is expected in the internal or external
+            # turbine library
+            if isinstance(t, str):
+                if t in turbine_definition_cache:
+                    continue
+
+                # Check if the file exists in the internal and/or external library
+                internal_fn = (default_turbine_library_path / t).with_suffix(".yaml")
+                external_fn = (self.turbine_library_path / t).with_suffix(".yaml")
+                in_internal = internal_fn.exists()
+                in_external = external_fn.exists()
+
+                # If an external library is used and there's a duplicate of an internal
+                # definition, then raise an error
+                is_unique_path = self.turbine_library_path != default_turbine_library_path
+                if is_unique_path and in_external and in_internal:
+                    raise ValueError(
+                        f"The turbine type: {t} exists in both the internal and external"
+                        " turbine library."
+                    )
+
+                if in_internal:
+                    full_path = internal_fn
+                elif in_external:
+                    full_path = external_fn
+                else:
+                    raise FileNotFoundError(
+                        f"The turbine type: {t} does not exist in either the internal or"
+                        " external turbine library."
+                    )
+                turbine_definition_cache[t] = load_yaml(full_path)
+
+        # Convert any dict entries in the turbine_type list to the type string. Since the
+        # definition is saved above, we can make the whole list consistent now to use it
+        # for mapping turbines later.
+        # We use a private variable here instead of self.turbine_type because self.turbine_type
+        # should always retain the input data. When this class is exported as_dict, the input
+        # types must be used. If we modify that directly and change its shape, recreating this
+        # class with a different layout but not a new self.turbine_type could cause the data
+        # to be out of sync.
+        _turbine_types = [
+            copy.deepcopy(t["turbine_type"]) if isinstance(t, dict) else t
+            for t in self.turbine_type
+        ]
```

Comment:
How do you feel about inserting the following two lines here (after line 153), so that `turbine_types` always ends up being a list of strings after this method is done?
```
        # Replace turbine_type list with all strings version
        self.turbine_type = _turbine_types.copy()
```

Reply:
One quirk that this introduces is that `fi.reinitialize()` takes an optional argument `turbine_type`, which I think can be a list of strings or dicts (confirming now); and if the `Farm` attribute `turbine_type` ends up being a list of strings only, even when `fi.reinitialize()` is supplied with dicts, this _could_ be confusing to the user.

Reply:
We shouldn't modify `self.turbine_type`. Since it is used in the `as_dict` method to export the Farm object, it needs to remain as input. The reason is because if we modify it, then export Farm but don't overwrite this attribute and do change other things that depend on it like `layout_x`, the data is out of sync and there's no way to resolve it. That's why I instead created the new local / private variable.

Where do we need to use `Farm.turbine_type` other than in this context? In other words, what's the need to make the elements of that list consistent? I'm wondering if there's another way to achieve the same thing.

Reply:
Ah fair, got it. The reason (for me) is that the exported Farm object ends up having this sort of odd structure of having attribute `turbine_type` which is a list of dicts, with each dict containing a key `"turbine_type"`; and the `turbine_type` attribute is then identical to the `turbine_definitions` attribute. But that doesn't create an actual runtime issue anywhere, just take a little bit to understand if exporting the configuration to a dictionary, changing things, and reinstantiating with the altered dictionary. As an example, I had a farm with mutliple turbine types (and many of each turbine type), and I wanted to change the pP value for just one of the turbine types. It was doable, but a bit confusing with pP exists in both the `"turbine_type"` and `"turbine_definition"` keys of the farm dict (in the end, changing the pP value in the `"turbine_type"` dict worked and overwrote those in `"turbine_definition"`, but I basically just got lucky)

Reply:
UPDATE: @rafmudaf and I spoke about this, and I had this wrong: "turbine_definitions" is _not_ an available key in the "farm" subdictionary of an exported FLORIS dictionary---only "turbine_type" is---so there is no extra confusion created (past "turbine_type" possibly being a list of dicts, each containing a key "turbine_type"). The code should be left as is without adding the lines I suggested. @rafmudaf, feel free to mark this conversation as resolved.